network/src/peers_registry.rs

```diff
@@ -112,235 +113,283 @@ impl PeerConnection {
         }
     }
 
-    pub fn append_addresses(&mut self, addresses: Vec<Multiaddr>) {
-        for addr in addresses {
-            if !self.remote_addresses.contains(&addr) {
-                self.remote_addresses.push(addr);
-            }
-        }
-    }
-
     #[inline]
-    pub fn is_outgoing(&self) -> bool {
+    pub fn is_outbound(&self) -> bool {
         self.endpoint_role == Endpoint::Dialer
     }
 
     #[allow(dead_code)]
     #[inline]
-    pub fn is_incoming(&self) -> bool {
-        !self.is_outgoing()
+    pub fn is_inbound(&self) -> bool {
+        !self.is_outbound()
     }
 
     #[allow(dead_code)]
-    pub fn add_remote_address(&mut self, remote_address: Multiaddr) {
-        if self
-            .remote_addresses
-            .iter()
-            .all(|addr| addr != &remote_address)
-        {
-            self.remote_addresses.push(remote_address);
-        }
+    #[inline]
+    fn network_group(&self) -> Group {
+        self.connected_addr.network_group()
     }
 }
 
 pub struct ConnectionStatus {
     pub total: u32,
-    pub unreserved_incoming: u32,
-    pub unreserved_outgoing: u32,
-    pub max_incoming: u32,
-    pub max_outgoing: u32,
+    pub unreserved_inbound: u32,
+    pub unreserved_outbound: u32,
+    pub max_inbound: u32,
+    pub max_outbound: u32,
 }
 
 pub(crate) struct PeersRegistry {
     // store all known peers
     peer_store: Arc<RwLock<Box<PeerStore>>>,
-    peer_connections: PeerConnections,
-    // max incoming limitation
-    max_incoming: u32,
-    // max outgoing limitation
-    max_outgoing: u32,
+    peers: PeerConnections,
+    // max inbound limitation
+    max_inbound: u32,
+    // max outbound limitation
+    max_outbound: u32,
     // Only reserved peers or allow all peers.
     reserved_only: bool,
-    deny_list: PeersDenyList,
+    reserved_peers: FnvHashSet<PeerId>,
+}
+
+fn find_most_peers_in_same_network_group<'a>(
+    peers: impl Iterator<Item = (&'a PeerId, &'a PeerConnection)>,
+) -> Vec<&'a PeerId> {
+    let mut groups: FnvHashMap<Group, Vec<&'a PeerId>> =
+        FnvHashMap::with_capacity_and_hasher(16, Default::default());
+    let mut largest_group_len = 0;
+    let mut largest_group: Group = Default::default();
+
+    for (peer_id, peer) in peers {
+        let group_name = peer.network_group();
+        let group = groups.entry(group_name.clone()).or_insert_with(Vec::new);
+        group.push(peer_id);
+        if group.len() > largest_group_len {
```

Comment:
we can use `max_by` to simplify code here

Reply:
Seem can't simplify code here, use `max_by` can remove only two lines but introduce an unnecessary implied loop.

Reply:
we can avoid extra variable allocating and simplify code to one line by using `max_by`:
```
peers.fold(FnvHashMap::default(), |mut acc, (peer_id, peer)| {
        acc.entry(peer.network_group()).or_insert_with(Vec::new).push(peer_id);
        acc
    }).values().max_by_key(|peers| peers.len())
```

Reply:
Cool! updated